sdks/java/core/src/main/java/org/apache/beam/sdk/coders/CoderProviders.java

```diff
@@ -178,7 +178,9 @@ public CoderProviderForCoder(TypeDescriptor<?> type, Coder<?> coder) {
     @Override
     public <T> Coder<T> coderFor(TypeDescriptor<T> type, List<? extends Coder<?>> componentCoders)
         throws CannotProvideCoderException {
-      if (!this.type.equals(type)) {
+      if (!(this.type.equals(type)
+          || (type.getRawType().getName().contains("AutoValue_")
+              && this.type.getRawType().isAssignableFrom(type.getRawType())))) {
```

Comment:
nit this can be a bit hard to understand. maybe we can add some comments or use variables to communicate the intent

```
boolean isTypeEqual = this.type.equals(type);
boolean isAutoValueConcrete = type.getRawType().getName().contains("AutoValue_") 
                                && this.type.getRawType().isAssignableFrom(type.getRawType());

if (!isTypeEqual && !isAutoValueConcrete) {
   throw
}
```

Reply:
Done